jOOQ-codegen/src/main/java/org/jooq/util/GenerationUtil.java

```diff
@@ -175,26 +258,38 @@ public static String convertToIdentifier(String literal, Language language) {
         StringBuilder sb = new StringBuilder();
 
         if ("".equals(literal))
-            return "_";
+            if (language == SCALA)
+                return "`_`";
+            else
+                return "_";
 
         for (int i = 0; i < literal.length(); i++) {
             char c = literal.charAt(i);
 
-            if (!Character.isJavaIdentifierPart(c))
-
-                // [#5424] Scala setters, by convention, end in "property_="
-                if (language == SCALA && c == '=' && i == literal.length() - 1)
-                    sb.append(c);
-                else
-                    sb.append(escape(c));
-            else if (i == 0 && !Character.isJavaIdentifierStart(literal.charAt(0)))
-                sb.append("_")
-                  .append(c);
+            // [#5424] Scala setters, by convention, end in "property_=", where "=" is an operator and "_" precedes it
+            if (language == SCALA && i == literal.length() - 1 && literal.length() >= 2 && literal.charAt(i - 1) == '_' && isScalaOperator(c))
+                sb.append(c);
+            else if (language == SCALA && !isScalaIdentifierPart(c))
+                sb.append(escape(c));
+            else if (language == JAVA && !Character.isJavaIdentifierPart(c))
+                sb.append(escape(c));
+            else if (language == SCALA && i == 0 && !isScalaIdentifierStart(c))
+                sb.append("_").append(c);
+            else if (language == JAVA && i == 0 && !Character.isJavaIdentifierStart(c))
+                sb.append("_").append(c);
             else
                 sb.append(c);
         }
 
-        return sb.toString();
+        String ret = sb.toString();
+
+        // In scala, _ should not be at the end of the identifier incase you have def foo_: Foo, so escape to def `foo_`: Foo
+        if (language == SCALA && ret.charAt(ret.length() - 1) == '_') {
+            return "`" + ret + "`";
+        } else {
+            return ret;
+        }
```

Comment:
Agh. This also causes regressions. OK, I'm not going to evaluate why they happened :) I guess this whole escaping Scala identifier stuff is just a huge can of worms. Great to have learned this lesson. In the future, I'll always opt for less idiomatic style in favour of going out of the way of these Scala language caveats